### PR TITLE
Process iterator in RPC result

### PR DESCRIPTION
**Pull request by shargon**
Alternative to https://github.com/neo-project/neo-modules/pull/577

### CODE REVIEW

**Comment by erikzhang** [comment]
If you return the values of the iterator directly, the developer will not be able to distinguish whether the method returns an array or an iterator.

**Comment by shargon** [comment]
> If you return the values of the iterator directly, the developer will not be able to distinguish whether the method returns an array or an iterator.

I think that api users doesn't care about the low level type, they want the value. Otherwise you will need to know first that it's an iterator result in order to make a different call.

**Comment by erikzhang** [comment]
At least we should not modify its type.

```
{
  "type": "InteropInterface",
  "value": []
}
```

**Comment by satoshichou** [comment]
I tried modules, making everything to config is not a good way, and server can't modify config when it's running a node. I find more and more paramaters are added to config.json, it's too complex for reading and understanding for a new developer. I like the style of https://github.com/neo-project/neo-modules/pull/577 much more.

**Comment by superboyiii** [comment]
As @satoshichou said, I totally agree. Don't make config too heavy and user could choose how much times iterator could run.

**Comment by steven1227** [comment]
I agree with @SueNEO  and @satoshichou  and I prefer that [Pull Request](https://github.com/neo-project/neo-modules/pull/577)

**Comment by erikzhang** [comment]
I support @shargon's solution. Because if you use index, it may cause data inconsistency. For example, if I call `invokeiterator` twice, the first request is 1~100, and the second request is 101~200. But if a new database insertion or deletion occurs between the two requests, the data returned between these two requests is inconsistent.

**Comment by nicolegys** [comment]
I'm confused about `MaxResultItems`. My test shows that when I set the value 10, if the amount of return result is more than 10, I will get this returned `"stack": "error: invalid operation"`. I think it should return the first 10 result and give a message, to tell the user that there are still some data not showed.

File: src/RpcServer/RpcServer.SmartContract.cs [hunk 1]
```diff
@@ -89,6 +92,34 @@ private JObject GetInvokeResult(byte[] script, Signers signers = null)
...
+                        max--;
+                        if (max <= 0) throw new InvalidOperationException();
+                    }
+
+                    JObject ret = new();
+                    ret["type"] = "InteropInterface";
+                    ret["value"] = ToJson(resultList, ref max);
+                    return ret;
+                }
+            }
```
Comment (satoshichou) [inline src/RpcServer/RpcServer.SmartContract.cs:118]: At least should show some warning and return current `InteropInterface` value.
```suggestion
string warning = null;
            if (max <= 0) throw new InvalidOperationException();
            max--;

            if (p is InteropInterface result)
            {
                var iterator = result.GetInterface<IIterator>();
                if (iterator != null)
                {
                    VM.Types.Array resultList = new();
                    while (iterator.Next())
                    {
                        resultList.Add(iterator.Value());
                        max--;
                        if (max <= 0)
                        {
                            max++;
                            warning = "MaxIteratorItems is acheived. Can't show all values.";
                            break;
                        }
                    }

                    JObject ret = new();
                    ret["type"] = "InteropInterface";
                    ret["value"] = ToJson(resultList, ref max);
                    ret["warning"] = warning;
                    return ret;
                }
            }
```

File: src/RpcServer/RpcServer.SmartContract.cs [hunk 2]
```diff
@@ -89,6 +92,41 @@ private JObject GetInvokeResult(byte[] script, Signers signers = null)
...
+                {
+                    VM.Types.Array resultList = new();
+                    while (iterator.Next())
+                    {
+                        resultList.Add(iterator.Value());
+                        max--;
+                        if (max <= 0)
+                        {
+                            max++;
+                            warning = "MaxIteratorItems is acheived. Can't show all values.";
```
Comment (shargon) [inline src/RpcServer/RpcServer.SmartContract.cs:114]:
```suggestion
                            warning = "MaxIteratorItems is achieved. The values was truncated.";
```

**Comment by nicolegys** [comment]
There's a boundary case. If the value of `MaxResultItems` is just enough to iterate over the data, the return result still shows the warning.

**Review by erikzhang — DISMISSED** [review]

File: src/RpcServer/RpcServer.SmartContract.cs [hunk 3]
```diff
@@ -78,13 +78,7 @@ private JObject GetInvokeResult(byte[] script, Signers signers = null)
...
             try
             {
-                int max = settings.MaxResultItems;
-                string warning = null;
-                json["stack"] = new JArray(engine.ResultStack.Select(p => ToJson(p, ref max, ref warning)).Where(u => u is not null));
-                if (!string.IsNullOrEmpty(warning))
-                {
-                    json["warning"] = warning;
-                }
+                json["stack"] = new JArray(engine.ResultStack.Select(p => ToJson(p, settings.MaxResultItems)));
```
Comment (shargon) [inline src/RpcServer/RpcServer.SmartContract.cs:81]: Now is not a `MaxResultItems` it's a `MaxIteratorResultItems`

Comment (erikzhang) [reply src/RpcServer/RpcServer.SmartContract.cs:81]: Yes. We should rename it.

**Review by nicolegys — DISMISSED** [review]
Test pass, just need to rename the parameter.

**Comment by nicolegys** [comment]
Since this solution not supports index, when the amount of data is very large, does user need to run his own node to get the complete data?

**Comment by erikzhang** [comment]
> Since this solution not supports index, when the amount of data is very large, does user need to run his own node to get the complete data?

Yes. I think public nodes are not suitable for running high-load tasks.

**Review by erikzhang — APPROVED** [review]

**Review by nicolegys — APPROVED** [review]

**Pull request merged**